sksports/model/heart_rate.py

```diff
@@ -0,0 +1,164 @@
+    return heart_rate - _model(power, *params)
+
+
+def _model_residual_least_square(params, heart_rate, power):
+    """Least-square error of the residuals."""
+    return np.sum(_model_residual(params, heart_rate, power) ** 2)
+
+
+class HeartRateRegressor(BaseEstimator, RegressorMixin):
+    """Estimate the heart-rate from power data.
+
+    The model used is based on the formulation in [1]_.
+
+    Parameters
+    ----------
+    solver : string, optional
+        The solver to estimate the model parameters. Can be either 'leastsq' to
+        apply Levenberg-Marquardt or one of 'method' from
+        :func:`scipy.optimize.minimize`.
+
+    hr_rest : float, default=75
+        Heart-rate frequency of the cyclist at rest.
+
+    hr_max : float, default=200
+        Heart-rate frequency of the cyclist at maximum.
+
+    hr_slope : float, default=0.30
+        Slope considered if the model is linear. FIXME
+
+    hr_drift : float, default=3e-5
+        Attenuation of the heart-rate other time. FIXME
+
+    rate_growth : float, default=24
+        Growth rate of the exponential when the power increases.
+
+    rate_decay : float, default=30
+        Decay rate of the exponential when the power decreases.
+
+    Attributes
+    ----------
+    hr_rest_ : float
+        Fitted heart-rate frequency of the cyclist at rest.
+
+    hr_max_ : float
+        Fitted heart-rate frequency of the cyclist at maximum.
+
+    hr_slope_ : float
+        Fitted slope considered if the model is linear. FIXME
+
+    hr_drift_ : float
+        Fitted attenuation of the heart-rate other time. FIXME
+
+    rate_growth_ : float
+        Fitted growth rate of the exponential when the power increases.
+
+    rate_decay_ : float
+        Fitted decay rate of the exponential when the power decreases.
+
+    References
+    ----------
+    .. [1] de Smet, Dimitri, et al. "Heart rate modelling as a potential
+       physical fitness assessment for runners and cyclists." Workshop at ECML
+       & PKDD. 2016.
+
+    """
+
+    def __init__(self, solver='Nelder-Mead', hr_rest=75, hr_max=200,
+                 hr_slope=0.30, hr_drift=3e-5, rate_growth=24, rate_decay=30):
+        self.solver = solver
+        self.hr_rest = hr_rest
+        self.hr_max = hr_max
+        self.hr_slope = hr_slope
+        self.hr_drift = hr_drift
+        self.rate_growth = rate_growth
+        self.rate_decay = rate_decay
+
+    def fit(self, X, y):
+        """Compute the parameters model.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples,)
+            The power data.
+
+        y : array-like, shape (n_samples,)
+            The heart-rate data.
+
+        Returns
+        -------
+        self
+
+        """
+        X, y = check_X_y(X, y, ensure_2d=False)
+        params_init = [self.hr_rest, self.hr_max, self.hr_slope, self.hr_drift,
+                       self.rate_growth, self.rate_decay]
+
+        if self.solver == 'leastsq':
```

Comment:
Can you explain this if-else? To me it seems the flow for `self.solver == 'leastsq'` is redundant. I.e. why can `minimize` be used for least-squares?

Reply:
`'leastqs` is actually using Levenverg-Marquardt algorithm. I was just playing with the different way of minimizing the problem. I am not really sure what is the best solution there. We can of course fix the solver as well.

NB: Levenvenberg-Maquardt will use  the gradient will an algorithm as the simplex (Nelder-Mead) will not use it. So depending of the optimization problem and the convexity it would be more appropriate to use one or another.